packages/host/app/components/operator-mode/add-field-modal.gts

```diff
@@ -0,0 +1,346 @@
+    this.chooseCardTask.perform();
+  }
+
+  @action onFieldNameInput(value: string) {
+    this.fieldName = value;
+  }
+
+  @action onCardinalityChange(id: 'one' | 'many'): void {
+    this.cardinality = id;
+  }
+
+  private chooseCardTask = restartableTask(async () => {
+    let chosenCatalogEntry = await chooseCard<CatalogEntry>({
+      filter: {
+        any: [
+          {
+            type: catalogEntryRef,
+          },
+        ],
+      },
+    });
+
+    if (chosenCatalogEntry) {
+      this.chosenCatalogEntry = chosenCatalogEntry;
+
+      // This transforms relative module paths, such as "../person", to absolute ones -
+      // we need that absolute path to load realm info
+      this.fieldModuleURL = new URL(
+        chosenCatalogEntry.ref.module,
+        chosenCatalogEntry.id,
+      );
+
+      this.chosenCatalogEntryRefCard = await loadCard(chosenCatalogEntry.ref, {
+        loader: this.loaderService.loader,
+        relativeTo: new URL(chosenCatalogEntry.id),
+      });
+    }
+  });
+
+  @action saveField() {
+    if (!this.chosenCatalogEntry || !this.args.card || !this.fieldName) {
+      throw new Error(
+        'bug: cannot save field without a selected card and a name',
+      );
+    }
+
+    let isField = this.chosenCatalogEntry.isField;
+
+    let relationshipType: FieldType;
+
+    if (isField) {
+      relationshipType =
+        this.cardinality === 'one' ? 'contains' : 'containsMany';
+    } else {
+      relationshipType = this.cardinality === 'one' ? 'linksTo' : 'linksToMany';
+    }
+
+    this.args.moduleSyntax.addField(
+      {
+        type: 'exportedName',
+        name: (
+          identifyCard(this.args.card)! as { module: string; name: string }
+        ).name,
+      },
+      this.fieldName,
+      this.chosenCatalogEntry.ref,
+      relationshipType,
+      new URL(this.chosenCatalogEntry.id),
+      this.loaderService.loader.reverseResolution(
+        makeResolvedURL(this.operatorModeStateService.state.codePath!).href,
+      ),
+      new URL(this.args.file.realmURL),
+    );
+
+    this.writeTask.perform(this.args.moduleSyntax.code());
+  }
+
+  get nameErrorMessage() {
+    if (this.fieldName) {
+      if (/\s/g.test(this.fieldName)) {
+        return 'Field names cannot contain spaces';
+      }
+
+      if (this.fieldName[0] === this.fieldName[0].toUpperCase()) {
+        return 'Field names must start with a lowercase letter';
+      }
+    }
+
+    return undefined;
+  }
+
+  get submitDisabled(): boolean {
+    return bool(
+      !this.fieldName ||
+        !this.chosenCatalogEntry ||
+        !this.chosenCatalogEntryRefCard ||
+        this.nameErrorMessage,
+    );
+  }
+
+  private writeTask = restartableTask(async (src: string) => {
+    // note that this write will cause the component to rerender, so
+    // any code after this write will not be executed since the component will
+    // get torn down before subsequent code can execute
+
+    this.args.file.write(src, true);
```

Comment:
Why is this a task?

Reply:
my guess is that args.file.write is async. so putting this in a task is probably a good idea since the component could be destroyed during the await. however there isn't an `await` which is probably going to leak async...

Reply:
Actually `write` is sync:

https://github.com/cardstack/boxel/blob/c7fe0eb5de735bb32ee06d5fb639dce59a9a83a5/packages/host/app/resources/file.ts#L183-L185

I suppose we could return the task promise and `await` for it in the add field modal code.

Reply:
sure, we might want a busy indicator while the save is happening (module updates would invalidate a lot more in the index than instance updates, so it's reasonable to expect that it might be slower to update).

one thing to keep in mind when you `await` a task `perform()`, is that you almost always want to `await` it within another task. if you `await` a task perform()` outside of a task, then the task is downcast to a normal promise and you lose all the cool features of having it be a task like component lifecycle awareness. In this case it's fine--just wanna make sure that i plant the seed for task `await` pitfalls

Reply:
> if you await a task perform() outside of a task, then the task is downcast to a normal promise and you lose all the cool features of having it be a task like component lifecycle awareness

great to know, thanks!

Reply:
I returned a promise from `write`and added a loading indicator on the button